deno/lib/helpers/util.ts

```diff
@@ -70,6 +70,24 @@ export namespace util {
   export type flatten<T> = identity<{ [k in keyof T]: T[k] }>;
   export type noUndefined<T> = T extends undefined ? never : T;
 
+  export function isEqual(a: any, b: any): boolean {
+    if (a === b) {
+      return true;
+    }
+
+    if (typeof a === "object" && typeof b === "object") {
+      const aKeys = objectKeys(a);
+      const bKeys = objectKeys(b);
+
+      if (aKeys.length !== bKeys.length) {
+        return false;
+      }
+
+      return aKeys.every((key) => isEqual(a[key], b[key]));
```

Comment:
This check doesn't guarantee that objects `a` and `b` are equal. For example:
```
isEqual({a: 1, b: undefined}, {a: 1, c: 1}) //=> true
```

Reply:
I didn't saw that possibility...
I added a check for existance of the property in b object with hasOwnProperty in a new commit. @webNeat Do you see other cases where this check would not be enough to check equality between a and b objects ?

Reply:
The only other case I can see with normal values is the case of `NaN`
```
const a = NaN
console.log(a === a) //=> false
```

Reply:
I think we can leave the case of NaN Like it is, for objects with NaN values in it will return false but as NaN === NaN returns false too this behavior is more consistant than returning true (I think returning a different result than the === operator for non object values would be inconsistant with the goal that function which is to provider an === operator with a consistant behavior of === check on values of objects instead of their memory adress).

So if you don't have any more problematic cases I mark this thread as resolved.

Reply:
I understand your point, if the goal is the provide an operator that behaves exactly the same as `===` then I agree with you.
if in the other hand, the goal is to use this to remove duplicated values, then two instances of `NaN` should be considered the same value, this is what the native `Set` does:
```ts
new Set([NaN, 1, NaN]) //=> Set(2) { NaN, 1 }
```
I have also checked the implementation of `eq` method from lodash and it seems to use the same logic of the `Set`: https://github.com/lodash/lodash/blob/master/test/eq.test.js#L9
```ts
_.eq(NaN, NaN) //=> true
```
What do you think?

Reply:
I'm not sure of that, the symbolic of NaN is to express a non numeric value that you tried to use as a number. But this is used no matter what non numeric value you have then it make sense that NaN != NaN as if you try to test 2 * x === 3 * x except for x = 0 you should always have false. If NaN === NaN you would here have 2 * x === 3 * x returning true for x = 'a' for exemple. So for me it makes sense returning false on isEqual(NaN, NaN) even it makes the array having multiple NaN. 

But I also understand your point of vue and honestly I can't say which one is the best point of vue.
Maybe @colinhacks as the owner of Zod could choose which behavior makes the more sense for him ?

Reply:
I would add that for me this point is not really a crucial concern has using Zod  with typescript strict type checking, you should never see your code generate NaN values and so this choice should quite never have any consequence.